torchrl/envs/libs/vmas.py

```diff
@@ -318,71 +467,93 @@ def _reset(
         )
         dones = self.read_done(dones)
 
-        agent_tds = []
-        for i in range(self.n_agents):
-            agent_obs = self.read_obs(obs[i])
-            agent_info = self.read_info(infos[i])
+        source = {"done": dones, "terminated": dones.clone()}
+        for group, agent_names in self.group_map.items():
+            agent_tds = []
+            for agent_name in agent_names:
+                i = self.agent_names_to_indices_map[agent_name]
+
+                agent_obs = self.read_obs(obs[i])
+                agent_info = self.read_info(infos[i])
+                agent_td = TensorDict(
+                    source={
+                        "observation": agent_obs,
+                    },
+                    batch_size=self.batch_size,
+                    device=self.device,
+                )
+                if agent_info is not None:
+                    agent_td.set("info", agent_info)
+                agent_tds.append(agent_td)
 
-            agent_td = TensorDict(
-                source={
-                    "observation": agent_obs,
-                },
-                batch_size=self.batch_size,
-                device=self.device,
-            )
-            if agent_info is not None:
-                agent_td.set("info", agent_info)
-            agent_tds.append(agent_td)
+            agent_tds = torch.stack(agent_tds, dim=1)
+            if not self.het_specs_map[group]:
+                agent_tds = agent_tds.to_tensordict()
+            source.update({group: agent_tds})
 
-        agent_tds = torch.stack(agent_tds, dim=1)
-        if not self.het_specs:
-            agent_tds = agent_tds.to_tensordict()
         tensordict_out = TensorDict(
-            source={"agents": agent_tds, "done": dones, "terminated": dones.clone()},
+            source=source,
             batch_size=self.batch_size,
             device=self.device,
         )
-
         return tensordict_out
 
     def _step(
         self,
         tensordict: TensorDictBase,
     ) -> TensorDictBase:
-        action = tensordict.get(("agents", "action"))
-        action = self.read_action(action)
+        agent_indices = {}
+        action_list = []
+        n_agents = 0
+        for group, agent_names in self.group_map.items():
+            group_action = tensordict.get((group, "action"))
+            group_action_list = list(self.read_action(group_action, group=group))
+            agent_indices.update(
+                {
+                    self.agent_names_to_indices_map[agent_name]: i + n_agents
+                    for i, agent_name in enumerate(agent_names)
+                }
+            )
+            n_agents += len(agent_names)
+            action_list += group_action_list
+        action = [action_list[agent_indices[i]] for i in range(self.n_agents)]
```

Comment:
this will convert actions from group dicts to the list (ordered as it should)

This bit of logic is the focus of the new test